presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -129,6 +121,107 @@ public static BindableAggregationFunction generateAggregationBindableFunction(Cl
+                    else {
+                        genericImplementations.add(implementation);
+                    }
+
+                    genericSignature = genericSignature.isPresent() ? genericSignature : Optional.of(implementation.getSignature());
+                    // TODO validateSignature(signature, implementation.getSignature());
+                }
+            }
+        }
+
+        Signature aggregateSignature = genericSignature.orElseGet(() -> getOnlyElement(exactImplementations.build().keySet()));
+
+        return new BindableAggregationFunction(aggregateSignature,
+                        header,
+                        new AggregationImplementations(
+                                exactImplementations.build(),
+                                specializedImplementations.build(),
+                                genericImplementations.build()));
+    }
+
+    private static AggregationImplementation parseImplementation(Class<?> aggregationDefinition, AggregationHeader header, Class<?> stateClass, Method inputFunction, Method outputFunction)
+    {
+        AggregationFunction aggregationAnnotation = aggregationDefinition.getAnnotation(AggregationFunction.class);
+
+        List<LongVariableConstraint> longVariableConstraints = parseLongVariableConstraints(inputFunction);
+        List<TypeVariableConstraint> typeVariableConstraints = parseTypeVariableConstraints(inputFunction);
+        List<Class<?>> signatureArgumentsTypes = parseSignatureArgumentsTypes(inputFunction);
+
+        List<TypeSignature> inputTypes = getInputTypesSignatures(inputFunction);
+        TypeSignature outputType = TypeSignature.parseTypeSignature(outputFunction.getAnnotation(OutputFunction.class).value());
+
+        Signature signature = new Signature(
+                header.getName(),
+                FunctionKind.AGGREGATE,
+                typeVariableConstraints,
+                longVariableConstraints,
+                outputType,
+                inputTypes,
+                false);
+
+        return new AggregationImplementation(signature, aggregationDefinition, stateClass, inputFunction, outputFunction, signatureArgumentsTypes);
+    }
+
+    private static List<Class<?>> parseSignatureArgumentsTypes(Method inputFunction)
+    {
+        ImmutableList.Builder<Class<?>> builder = ImmutableList.builder();
+
+        int stateId = findAggregationStateParamId(inputFunction);
+
+        for (int i = 0; i < inputFunction.getParameterCount(); i++) {
+            Class<?> parameterType = inputFunction.getParameterTypes()[i];
+            Annotation[] annotations = inputFunction.getParameterAnnotations()[i];
+
+            // Skip injected parameters
+            if (parameterType == ConnectorSession.class) {
+                continue;
+            }
+
+            if (Arrays.stream(annotations).anyMatch(AggregationCompiler::isAggregationMetaAnnotation)) {
+                continue;
+            }
+
+            if (i == stateId) {
```

Comment:
This check is unnecessary because the parameter is marked with `@AggregationState`, and you would filter out these positions in the previous `if`.